File: complex-1.py
class Complex:
  def __init__(self, real=0.0, imag=0.0):
      self.real = real
      self.imag = imag

  def __add__(self, other):
      if isinstance(other, Complex):
          return Complex(self.real + other.real, self.imag + other.imag)
      return Complex(self.real + other, self.imag)

  def __radd__(self, other):
      return self + other

  def __sub__(self, other):
      if isinstance(other, Complex):
          return Complex(self.real - other.real, self.imag - other.imag)
      return Complex(self.real - other, self.imag)

  def __rsub__(self, other):
      return Complex(other - self.real, -self.imag)

  def __mul__(self, other):
      if isinstance(other, Complex):
          return Complex(self.real * other.real - self.imag * other.imag,
                         self.real * other.imag + self.imag * other.real)
      return Complex(self.real * other, self.imag * other)

  def __rmul__(self, other):
      return self * other

  def __truediv__(self, other):
      if isinstance(other, Complex):
          denom = other.real**2 + other.imag**2
          return Complex((self.real * other.real + self.imag * other.imag) / denom,
                         (self.imag * other.real - self.real * other.imag) / denom)
      return Complex(self.real / other, self.imag / other)

  def __rtruediv__(self, other):
      denom = self.real**2 + self.imag**2
      return Complex((other * self.real) / denom, (-other * self.imag) / denom)

  def __str__(self):
      return f"({self.real} {'+' if self.imag >= 0 else '-'} {abs(self.imag)}i)"


if __name__ == "__main__":
  a = Complex(1, 2)
  b = Complex(2, -1)
  i = 5

  print(f"{a} + {b} = {a + b}")
  print(f"{a} - {b} = {a - b}")
  print(f"{a} * {b} = {a * b}")
  print(f"{a} / {b} = {a / b}")

  print(f"{a} + {i} = {a + i}")
  print(f"{i} + {a} = {i + a}")
  print(f"{a} - {i} = {a - i}")
  print(f"{i} - {a} = {i - a}")
  print(f"{a} * {i} = {a * i}")
  print(f"{i} * {a} = {i * a}")
  print(f"{a} / {i} = {a / i}")
  print(f"{i} / {a} = {i / a}")

File: Complex-1.java
class Complex {

    private double real;
    private double imag;

    public Complex() {
        this(0.0, 0.0);
    }

    public Complex(double real) {
        this(real, 0.0);
    }

    public Complex(double real, double imag) {
        this.real = real;
        this.imag = imag;
    }

    public Complex add(Complex o) {
        return new Complex(this.real + o.real, this.imag + o.imag);
    }

    public Complex add(double d) {
        return new Complex(this.real + d, this.imag);
    }

    public Complex sub(Complex o) {
        return new Complex(this.real - o.real, this.imag - o.imag);
    }

    public Complex sub(double d) {
        return new Complex(this.real - d, this.imag);
    }

    public Complex mul(Complex o) {
        double r = this.real * o.real - this.imag * o.imag;
        double i = this.real * o.imag + this.imag * o.real;
        return new Complex(r, i);
    }

    public Complex mul(double d) {
        return new Complex(this.real * d, this.imag * d);
    }

    public Complex div(Complex o) {
        double denom = o.real * o.real + o.imag * o.imag;
        double r = (this.real * o.real + this.imag * o.imag) / denom;
        double i = (this.imag * o.real - this.real * o.imag) / denom;
        return new Complex(r, i);
    }

    public Complex div(double d) {
        return new Complex(this.real / d, this.imag / d);
    }

    public static Complex add(double d, Complex c) {
        return new Complex(d + c.real, c.imag);
    }

    public static Complex sub(double d, Complex c) {
        return new Complex(d - c.real, -c.imag);
    }

    public static Complex mul(double d, Complex c) {
        return new Complex(d * c.real, d * c.imag);
    }

    public static Complex div(double d, Complex c) {
        double denom = c.real * c.real + c.imag * c.imag;
        double r = (d * c.real) / denom;
        double i = (-d * c.imag) / denom;
        return new Complex(r, i);
    }

    @Override
    public String toString() {
        return "(" + real + (imag >= 0 ? " + " : " - ") + Math.abs(imag) + "i)";
    }
}


public class Main {
    public static void main(String[] args) {
        Complex a = new Complex(1, 2);
        Complex b = new Complex(2, -1);
        double d = 5.0;

        System.out.println(a + " + " + b + " = " + a.add(b));
        System.out.println(a + " - " + b + " = " + a.sub(b));
        System.out.println(a + " * " + b + " = " + a.mul(b));
        System.out.println(a + " / " + b + " = " + a.div(b));

        System.out.println(a + " + " + d + " = " + a.add(d));
        System.out.println(d + " + " + a + " = " + Complex.add(d, a));
        System.out.println(a + " - " + d + " = " + a.sub(d));
        System.out.println(d + " - " + a + " = " + Complex.sub(d, a));
        System.out.println(a + " * " + d + " = " + a.mul(d));
        System.out.println(d + " * " + a + " = " + Complex.mul(d, a));
        System.out.println(a + " / " + d + " = " + a.div(d));
        System.out.println(d + " / " + a + " = " + Complex.div(d, a));
    }
}
